### PR TITLE
[MIG] Migrate mis_builder to 16.0

### PR DESCRIPTION
**Pull request by dzungtran89**
- Migrated the widget `mis_report_widget` to owl
- Fix `render` style NUM

### CODE REVIEW

**Comment by dzungtran89** [comment]
@sbidoul could you please have a look on the PR?

File: mis_builder/wizard/mis_builder_dashboard.py [hunk 1]
```diff
@@ -71,11 +71,15 @@ def action_add_to_dashboard(self):
             arch = self.env["ir.ui.view.custom"].browse(last_customization[0].id).arch
         new_arch = etree.fromstring(arch)
         column = new_arch.xpath("//column")[0]
+        # Due to native dashboard doesn't support form view
+        # add "from_dashboard" to context to get correct views in "get_views"
+        context = self._context.copy()
+        context['from_dashboard'] = True
```
Comment (sbidoul) [inline mis_builder/wizard/mis_builder_dashboard.py]:
```suggestion
        context = dict(self.env.context, from_dashboard=True)
```

to avoid using the deprecated `self._context`?

File: mis_builder/views/mis_report.xml [hunk 2]
```diff
@@ -119,8 +122,8 @@
                     <field name='id' invisible='1' />
                     <field
                         name="report_id"
-                        invisible="1"
                         attrs="{'required': [('id', '!=', False)]}"
+                        options="{'no_create_edit': True, 'no_open': True}"
```
Comment (sbidoul) [inline mis_builder/views/mis_report.xml]: Is this change needed for the migration? If not can you do it as a separate PR to facilitate review?

File: mis_builder/views/mis_report.xml [hunk 3]
```diff
@@ -237,7 +240,7 @@
                                 </div>
                             </group>
                             <group>
-                                <div>
+                                <div colspan="2">
```
Comment (sbidoul) [inline mis_builder/views/mis_report.xml]: Is this change needed for the migration? If not can you do it as a separate PR so as to facilitate backports?
Cherry on the cake would be before/after screenshots to facilitate review :)

File: mis_builder/tests/test_mis_report_instance.py [hunk 4]
```diff
@@ -564,21 +564,6 @@ def test_multi_company_onchange(self):
...
-        )._compute_matrix()
-        for row in matrix.iter_rows():
-            vals = [c.val for c in row.iter_cells()]
-            if row.kpi.name == "k1":
-                self.assertEqual(vals, [AccountingNone, AccountingNone, AccountingNone])
-            elif row.kpi.name == "k2":
-                self.assertEqual(vals, [AccountingNone, AccountingNone, None])
-            elif row.kpi.name == "k4":
-                self.assertEqual(vals, [AccountingNone, AccountingNone, 1.0])
-
```
Comment (sbidoul) [inline mis_builder/tests/test_mis_report_instance.py]: Please keep the analytic filter tests even if the branch is red for now. We'll handle these separately.

File: mis_builder/__manifest__.py [hunk 5]
```diff
@@ -30,8 +30,9 @@
     "assets": {
         "web.assets_backend": [
             "mis_builder/static/src/css/custom.css",
-            "mis_builder/static/src/js/mis_report_widget.js",
-            "mis_builder/static/src/xml/mis_report_widget.xml",
```
Comment (sbidoul) [inline mis_builder/__manifest__.py:34]: Let's simply remove them if they are not used anymore?

**Review by sbidoul — COMMENTED** [review]
Thanks a lot for working on this! I made a few minor comments, then we can merge it.

**Comment by dzungtran89** [comment]
Hi @sbidoul , thanks for your review. The PR has been updated

File: mis_builder/views/mis_report.xml [hunk 6]
```diff
@@ -119,7 +121,6 @@
                     <field name='id' invisible='1' />
                     <field
                         name="report_id"
-                        invisible="1"
```
Comment (sbidoul) [inline mis_builder/views/mis_report.xml]: Why does this field need to become visible?

File: mis_builder/views/mis_report.xml [hunk 7]
```diff
@@ -102,12 +103,13 @@
             </form>
         </field>
     </record>
+
     <record id="mis_report_view_kpi_form" model="ir.ui.view">
         <field name="name">mis.report.view.kpi.form</field>
         <field name="model">mis.report.kpi</field>
         <field name="arch" type="xml">
             <form string="MIS Report KPI">
-                <group col="4">
```
Comment (sbidoul) [inline mis_builder/views/mis_report.xml]: A little doubt about this one. Can you show the before / after screenshot?

File: mis_builder/static/src/components/mis_report_widget.js [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    }
+    async drilldown (event) {
+        const drilldown = $(event.target).data("drilldown");
+        const action = await this.orm.call(
+            "mis.report.instance",
+            "drilldown",
+            [this._instanceId(), drilldown],
+            {context: this.user_context},
+        )
+        this.legacyActionManager.do_action(action);
```
Comment (sbidoul) [inline mis_builder/static/src/components/mis_report_widget.js]: Do you know why we need to use a "legacy" action manager here? Sorry for the maybe dumb question, I'm not very familiar with the frontend framework.

File: mis_builder/views/mis_report.xml (hunk 7 above)
Comment (dzungtran89) [reply mis_builder/views/mis_report.xml]: - Before with `col="4"`

image

- After with `colspan="4"`

image

The main diff between them are about the right padding of the fields, indeed it should not have been included in the PR in order to fascilate the review

File: mis_builder/static/src/components/mis_report_widget.js (hunk 8 above)
Comment (dzungtran89) [reply mis_builder/static/src/components/mis_report_widget.js]: Thanks for the question, it basically makes the legacy `do_action` work with the owl action service, you can see [this](https://github.com/odoo/odoo/blob/16.0/addons/web/static/src/legacy/backend_utils.js#L254) for references

This allows us to click on the report numbers for drilldown-ing the details

image

**Comment by dzungtran89** [comment]
(I pushed a fix for the button `refresh` on the widget)

File: mis_builder/static/src/components/mis_report_widget.js (hunk 8 above)
Comment (dzungtran89) [reply mis_builder/static/src/components/mis_report_widget.js]: anw because of your nice question, we can use this `this.action = useService("action");` instead indeed

File: mis_builder/static/src/components/mis_report_widget.xml [hunk 9]
```diff
@@ -47,6 +51,7 @@
                             <th
                                 t-foreach="row.cols"
                                 t-as="col"
+                                t-key="col_index"
```
Comment (sbidoul) [inline mis_builder/static/src/components/mis_report_widget.xml]: Is this col_index used anywhere?

File: mis_builder/static/src/components/mis_report_widget.xml [hunk 10]
```diff
@@ -59,24 +64,26 @@
                         </tr>
                     </thead>
                     <tbody>
-                        <tr t-foreach="widget.mis_report_data.body" t-as="row">
-                            <td t-att="{'style': row.style}">
-                                <t t-esc="row.label" />
-                                <t t-if="row.description">
+                        <tr t-foreach="state.mis_report_data.body" t-as="body_row" t-key="body_row_index">
```
Comment (sbidoul) [inline mis_builder/static/src/components/mis_report_widget.xml]: Any particular reason to rename row in body_row? If not let's keep the previous value to minimize the diff: this facilitates ports across versions.

File: mis_builder/static/src/components/mis_report_widget.js (hunk 8 above)
Comment (sbidoul) [reply mis_builder/static/src/components/mis_report_widget.js]: Great! Thanks for investigating.

File: mis_builder/views/mis_report.xml (hunk 7 above)
Comment (sbidoul) [reply mis_builder/views/mis_report.xml]: This looks better indeed. If this applies to other versions, don't hesitate to do a PR to any version, and I'll forward/backport.

File: mis_builder/static/src/components/mis_report_widget.xml (hunk 9 above)
Comment (anothingguy) [reply mis_builder/static/src/components/mis_report_widget.xml]: It won't be used anywhere, but in Odoo 16, OWL 2.0 required `t-key` tag for `t-foreach` ([ref](https://github.com/odoo/owl/blob/master/CHANGELOG.md#20-t-foreach-should-always-have-a-corresponding-t-key))

**Comment by dzungtran89** [comment]
thanks @anothingguy 
@sbidoul , thanks for asking, the PR has been updated

**Comment by sbidoul** [comment]
Thanks for this work, folks!
